web-api-validation-expression.md

```diff
@@ -1,98 +1,285 @@
+| **RESTRICT_PICKLIST** | `LIST(CHAR)` | Expressions of this type are designed to return one or more lookup values that are to be removed from the lookup list that is defined for the field the rule is being executed on. This is always the entire set of values to remove from the lookup. <br /><br />In other words, if this returns a blank list or `.EMPTY.`, the entire set of lookup values is to be displayed.<br /><br />The value of this expression MUST be a `LIST`, rather than `Exp`. <br /><br />All members of the list MUST be of the same data type as that of the field the rule is being executed on. |
+| **SET_PICKLIST** | `LIST(CHAR)` | Expressions of this type are designed to return one or more lookup values that are to be used in the lookup list defined for the field the rule is being executed on.<br /><br />The value of this expression MUST be a `LIST`, rather than `Exp`. <br /><br />Every member of the list MUST exist in the lookup list as defined in the metadata for the field the rule is being executed on. |
+| **SET_DISPLAY** | `BOOLEAN` | Expressions of this type are designed to allow a client to make fields visible or invisible based on the evaluation of an expression. The result of this expression has no effect on whether a field is READ ONLY or not. |
+
+### Section 2.1.2: Validation Expressions
+
+There are two files that define the Validation Expressions grammar:
+* [ANTLR 4 grammar](./artifacts/grammars/RCP-019/rcp019.g4) - Contains [non-terminal symbols](https://en.wikipedia.org/wiki/Terminal_and_nonterminal_symbols#Nonterminal_symbols)
+* [ANTLR 4 lexer](./artifacts/grammars//RCP-019/rcp019Lexer.g4) - Contains [terminal symbols](https://en.wikipedia.org/wiki/Terminal_and_nonterminal_symbols#Terminal_symbols)
+
+#### Basic Examples
+
+##### Arithmetic Operators
 `3 + 5`
 
 `1 * 3 + 2 - 5`
 
-### Bracketed RETSNAMEs  
-`[ListPrice] > 0`
+##### FIELD_NAME
+To get current value of any field, use the Data Dictionary name: `ListPrice`
 
-### RCP61 added support for unbracketed RETSNAMEs. This also works for DICTNAMEs.
-`ListPrice > 0`
+##### Bracketed FIELD_NAMEs  
+Supported for backwards compatibility with RETS: 
+`[ListPrice]`
 
-### To get current value of any field, just use the `DICTNAME` of that field  
-`ListPrice`
+##### Simple Comparisons
+`ListPrice > 0`
 
-### Access the last value of a field
+##### Access the Last Value of a Field
+`LAST Status != 'Sold'`
 
-`LAST Status != "Sold"`
+##### Field Change Detection
+Returns true if ListPrice has changed.
 
-### Field change detection
 `ListPrice != LAST ListPrice`
 
-### Function calls are also supported
+##### Function Calls
 `foo()`  
-`foo ('bar', 'baz', 42)`
+`foo('bar', 'baz', 42)`
 
-_Cannot use any RETSNAMEs, DICTNAMEs, or reserved words_
+_Functions MUST NOT use any FIELD_NAMEs or reserved words._
 
-### Lists of expressions
+##### Lists of Expressions
 `(ListPrice, Status, 3+5, LAST Status, (1, 2, 3))`

```

Comment:
I don't know if it's solvable anymore due to existing implementations, but this always bothered me. According to the grammar...

| Input | Type  |
| :---- | :---- |
| `(1, 2, 3, 4, 5)` | List of 5 items |
| `(1, 2, 3, 4)`    | List of 4 items |
| `(1, 2, 3)`       | List of 3 items |
| `(1, 2)`          | List of 2 items |
| `(1)`             | ⚠️ ***A NUMBER*** ⚠️  |
| `()`              | List of 0 items |

Where this has bitten me when authoring rules

| Expression | Validity |
| :---- | :---- |
| `Status .IN. ('Active', 'Closed', 'Pending')` | ✅ |
| `Status .IN. ('Active', 'Closed')` | ✅ |
| `Status .IN. ('Active')` | ❌ `.IN.` operator isn't defined for CHAR .IN. CHAR |
| `Status .IN. ()` | ✅ |

Fortunately, always using the `SET` function is consistent.

| Expression | Validity |
| :---- | :---- |
| `Status .IN. SET('Active', 'Closed', 'Pending')` | ✅ |
| `Status .IN. SET('Active', 'Closed')` | ✅ |
| `Status .IN. SET('Active')` | ✅ |
| `Status .IN. SET()` | ✅ |

But also, we can't redefine `(exp)` to be a one-element list, because `(exp)` is required for precedence in situations like `(FirstYN .AND. SecondYN) .OR. ThirdYN`.

Reply:
I commented on this in the original grammar ... 

https://github.com/darnjo/rcp019/blob/5b77e7e8860b044a65b2611f829c1240d60ca3e3/rcp019.g4#L59-L60

This was left in place since the original grammar defined it. Use `LIST()` and `SET()` instead for more normal list behavior.